Add examples/ReinterpretLog.hs

mitchellwrosen opened the pull request:
Hi, today I was playing around with `fused-effects` and attempted to define an effect for structured logging. It was a bit difficult to figure out so I decided to work it into an example for others.

At a high level:

- Define an effect for logging an arbitrary `a`
- Define a log carrier for printing logged `String`s
- Define a log carrier for reinterpreting `a`s as `String`s

This example is not yet structured like the others (no spec) because I'm not exactly sure this is the recommended way of doing this, so comments on how the code works are very welcome!

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+--
+-- * Next, we will define a logging carrier that prints strings to stdout.
+--
+-- * Finally, we will bridge the two with an effect carrier that reinterpretes
```
```suggestion
-- * Finally, we will bridge the two with an effect carrier that reinterprets
```

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+-- Log an 'a', then continue with 'k'.
+data Log :: Type -> (Type -> Type) -> Type -> Type where
+  Log :: a -> k -> Log a m k
+  deriving stock (Functor)
```
Would it be better (for the purpose of pedagogy) to write this without GADT syntax? Just thinking that this requires readers to already understand one more thing than they would otherwise have to.

mitchellwrosen replied:
Will do!

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+
+instance
+     (Carrier sig m, MonadIO m)
+  => Carrier (Log String :+: sig) (LogStdoutC m) where
```
This is a superb example of a carrier for a specialization of an effect type ✨

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+  = ReinterpretLogC ((s -> t) -> m a)
+  deriving stock (Functor)
+  deriving (Applicative, Monad, MonadIO)
+       via (ReaderT (s -> t) m)
```
We currently support back to ghc 8.2, so `-XDerivingVia` won’t fly.

I’d suggest instead defining `ReinterpretLogC s t m a` as a wrapper around `ReaderC (s -> t) m a` and using `-XGeneralizedNewtypeDeriving`.

mitchellwrosen replied:
Sounds good.

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+    L (Log s k) ->
+      ReinterpretLogC $ \f -> do
+        log (f s)
+        reinterpretLog f k
```
If `ReinterpretLogC` is defined as a wrapper around `ReaderC`, its `Carrier` instance can be defined in terms of `Reader`, e.g. by using `ask` to retrieve the function.

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+
+    R other ->
+      ReinterpretLogC $ \f ->
+        eff (handlePure (reinterpretLog f) other)
```
Likewise, wrapping `ReaderC` simplifies this case to `eff (handleCoercible other)`.

mitchellwrosen replied:
Nice. I believe you need an `R` in there as well.

robrix replied:
Whoops! Indeed you do, good catch 😊

robrix commented on examples/ReinterpretLog.hs:
```diff
@@ -0,0 +1,159 @@
+
+--------------------------------------------------------------------------------
+-- The logging effect carriers
+--------------------------------------------------------------------------------
```
In order to define a spec to show off how this operates, I’d suggest defining a carrier which simply collects the logged messages (of whatever type), and then testing the list of collected (transformed) messages against the expected strings.

mitchellwrosen commented:
@robrix Thanks for the feedback, I believe I've addressed your comments. I also added some more comments in the code, but may have gone overboard... so please re-review if you'd like!

patrickt commented:
This is looking good to me once @robrix gives the okay.

robrix commented:
Merging! Thanks so much @mitchellwrosen! 💖
